tcpdf.php

```diff
@@ -13675,7 +13680,127 @@ protected function applyTSA($signature) {
+		$hash = hash('sha1', hex2bin($pkcs7EncryptedDigest));
+		
+		// Build timestamp request data
+		$tsReqData = seq(
+										int(1).
+										seq(
+											seq(
+												"06052B0E03021A". // Obj_sha1
+												"0500" // Null
+												).
+											oct($hash)
+											).
+										int(hash('crc32', rand())).	// Add random nonce request
+										'0101ff' // set certReq true to tell TSA server to include SigningCertificate
+										);
+		
+		$raw_data = hex2bin($tsReqData);
+
+		//Send request to TSA Server with Curl
+		if(extension_loaded('curl')) {
+			$ch = curl_init();
+			curl_setopt($ch, CURLOPT_URL, $this->tsa_data['tsa_host']);
+			curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+			curl_setopt($ch, CURLOPT_POST, 1);
+			curl_setopt($ch, CURLOPT_HTTPHEADER, array(
+														'Content-Type: application/timestamp-query',
+														'User-Agent: TCPDF'
+														)
+						);
+			curl_setopt($ch, CURLOPT_POSTFIELDS, $raw_data);
+
+			
+			// can't send tsRequest, Timestamp failed!
+			if(!$tsResponse = curl_exec($ch)) {
+				return $signature;
+			}
+			
+			// parse timestamp response data
+			$hexTsaResponse = bin2hex($tsResponse);
+			if(!$parseTimeStampResp = asn1parse($hexTsaResponse)) { // bad TSA Reponse
+				return $signature;
+			}
+			
+			// verify tsa response PKIStatusInfo and TimeStampToken exists
+			if(!$TimeStampResp = asn1parse($parseTimeStampResp[0][1])) {
+				return $signature;
+			}
+	
+			// Select timeStampToken only. must ignore response status data (in first sequence if exist, select 2nd sequence)
+			if(count($TimeStampResp) > 1) {
+				$TSTInfo = $TimeStampResp[1][1];	// TSTInfo
+			} else if (count($TimeStampResp) == 1) {
+				$TSTInfo = $TimeStampResp[0][1];	// TSTInfo
+			} else { // TimeStampResp not containts 1 or 2 fields
+				return $signature;
+			}
+
+			// Add timestamp in TCPDF Signature
+			// Create timestamp pkcs#7 data
+			$TimeStampToken = seq(
+								"060B2A864886F70D010910020E".	// OBJ_id_smime_aa_timeStampToken
+								set(
+									seq(
+										$TSTInfo // TSTInfo
+										)
+									)	
+								);
+			
+			$time = seq(
+						$pkcs7signerInfos[0][1].
+						explicit(1,
+								$TimeStampToken
+								)
+						);
+			
+			$pkcs7contentSignedData=seq(
+										int(1).	// version
+										set($pkcs7SignedData[1][1]).	// digestAlgorithms
+										seq($pkcs7SignedData[2][1]).	// contentInfo
+										explicit(0,
+												$pkcs7SignedData[3][1]
+												).	// certificates [0] IMPLICIT ExtendedCertificatesAndCertificates
+										set(
+											$time
+											)
+										);
+			$pkcs7ContentInfo = seq(
+									"06092A864886F70D010702".	//	ContentType OBJ_pkcs7_signed
+									explicit(0,($pkcs7contentSignedData))	//	content
+									).
+									// "0000"; // sometime needed for backward compatibility
+									"";
+			
+			$signature = $pkcs7ContentInfo;
+		}
```

Comment:
can you indent this with spaces and move all this code into a protected function you would call here to fetch the signature ?
That would avoid adding that much code into this function

Also, I think you could keep the actual value of signature_max_length and only increase it (if the value < what you need) when calling the signature process ?
Would that keep other pdf documents the same as before (the ones without a signature)

Reply:
i think signature_max_length value have to be reserved before signing process, thats why it not set to 0. will padded by zero after signing. this value affect only signed file size to several kilobytes. Tried to lower this value around 14k and still done with apple tsa.
want move to object oriented but not at the moment. maybe to closed this pull request while i'll periodically update script.

Reply:
Please do not close this pull request
If you invite me on your repository I may clean up the implementation for you

Reply:
ok.

Reply:
Can I start pushing work or do you have more to push first?

Reply:
wait for awhile. i'm updating code to classes.

Reply:
okay, let me know

Reply:
I have done, please review and make some corrections if needed.
I've minimized the script, considering most of the general purpose asn.1 encoders/decoders use complex libraries.